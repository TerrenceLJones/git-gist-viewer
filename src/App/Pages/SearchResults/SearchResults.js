import React from 'react';

function makeSearchItems(item) {
    const imgStyles = {
        borderRadius: '50px',
        height: '50px',
        width: '50px'
    };

    const itemStyles = {
        height: '100px',
        margin: '20px',
        width: '100px'
    };

    return (
        <li key={item.id} id={item.login} style={itemStyles}>
            <img style={imgStyles} src={`${item.avatar_url}`} alt="" />
            <p>{item.login}</p>
        </li>
    );
}

const SearchResults = (props) => {
    const { results, handleSearchItemClick } = props
    const listStyles = {
        listStyle: 'none',
        display: 'flex',
        flexDirection: 'row',
        flexWrap: 'wrap'
    };

    return (
        <ul style={listStyles} onClick={handleSearchItemClick}>
            {results.map(makeSearchItems)}
        </ul>
    );
}
export default SearchResults;

// back button doesn't update search input
