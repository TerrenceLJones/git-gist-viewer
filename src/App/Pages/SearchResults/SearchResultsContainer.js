import React from 'react';
import { compose, lifecycle, withHandlers } from 'recompose';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { parse } from 'query-string'
import _ from 'lodash';

import { nonOptimalStates } from 'redux/utils/nonOptimalStates';
import { apiSelectors } from 'redux/modules/api'
import { usersActions, usersTypes } from 'redux/modules/users';
import SearchResults from './SearchResults';

const resultsLoadingSelector = apiSelectors.createLoadingSelector([usersTypes.API_HANDLER_SEARCH_USERS]);
const resultsErrorSelector = apiSelectors.createErrorMessageSelector([usersTypes.API_HANDLER_SEARCH_USERS]);

const LoadingDisplay = () => "Loading...";
const ErrorDisplay = ({ error }) => error;
const NoResultsDisplay = () => <span>No Results. Please try your search again.</span>;
const SearchResultsContainer = ({ handleSearchItemClick, results }) => {
    return <SearchResults results={results} handleSearchItemClick={handleSearchItemClick} />
};

const mapStateToProps = (state, ownProps) => {
    const queryString = _.get(ownProps, 'location.search');
    const searchValue = _.get(parse(queryString), 'q')
    const results = state.users.searchResults;

    return {
        isLoading: results.length ? resultsLoadingSelector(state) : true,
        loadError: resultsErrorSelector(state),
        results,
        searchValue
    }
};

const mapDispatchToProps = dispatch => ({
    searchUsers: queryString => dispatch(usersActions.searchUsersRequest(queryString))
});

const enhance = compose(
    connect(mapStateToProps, mapDispatchToProps),
    withRouter,
    lifecycle({
        componentDidMount() {
            this.props.searchUsers(this.props.searchValue);
        },
        componentDidUpdate(prevProps) {
            const prevSerchValue = prevProps.searchValue;
            const currSearchValue = this.props.searchValue;

            if (currSearchValue !== prevSerchValue) {
                this.props.searchUsers(currSearchValue);
            }
        }
    }),
    withHandlers({
        handleSearchItemClick: ({ history }) => ({ target }) => {
            const userName = target.parentElement.id;
            history.push(`/users/${userName}`);
        }
    }),
    nonOptimalStates([
        { when: ({ loadError }) => !!loadError, render: ErrorDisplay },
        { when: ({ results }) => _.isEmpty(results), render: NoResultsDisplay },
        { when: ({ isLoading }) => isLoading, render: LoadingDisplay }
    ])
);

export default enhance(SearchResultsContainer);
