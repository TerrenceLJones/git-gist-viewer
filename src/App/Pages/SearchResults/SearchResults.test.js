import React from 'react';
import { shallow } from 'enzyme';
import SearchResults from './SearchResults';

it('renders without crashing', () => {
    const results = [{ id: 1, id: 2 }];
    expect(shallow(<SearchResults results={results} />)).toMatchSnapshot();
});
