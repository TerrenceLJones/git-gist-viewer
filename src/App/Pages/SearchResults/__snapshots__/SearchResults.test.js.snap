// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchResults
    results={
        Array [
            Object {
              "id": 2,
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li
          id={undefined}
          style={
                    Object {
                              "height": "100px",
                              "margin": "20px",
                              "width": "100px",
                            }
          }
>
          <img
                    alt=""
                    src="undefined"
                    style={
                              Object {
                                        "borderRadius": "50px",
                                        "height": "50px",
                                        "width": "50px",
                                      }
                    }
          />
          <p />
</li>,
      ],
      "onClick": undefined,
      "style": Object {
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "listStyle": "none",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt=""
              src="undefined"
              style={
                            Object {
                                          "borderRadius": "50px",
                                          "height": "50px",
                                          "width": "50px",
                                        }
              }
/>,
            <p />,
          ],
          "id": undefined,
          "style": Object {
            "height": "100px",
            "margin": "20px",
            "width": "100px",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "",
              "src": "undefined",
              "style": Object {
                "borderRadius": "50px",
                "height": "50px",
                "width": "50px",
              },
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "p",
          },
        ],
        "type": "li",
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li
            id={undefined}
            style={
                        Object {
                                    "height": "100px",
                                    "margin": "20px",
                                    "width": "100px",
                                  }
            }
>
            <img
                        alt=""
                        src="undefined"
                        style={
                                    Object {
                                                "borderRadius": "50px",
                                                "height": "50px",
                                                "width": "50px",
                                              }
                        }
            />
            <p />
</li>,
        ],
        "onClick": undefined,
        "style": Object {
          "display": "flex",
          "flexDirection": "row",
          "flexWrap": "wrap",
          "listStyle": "none",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt=""
                src="undefined"
                style={
                                Object {
                                                "borderRadius": "50px",
                                                "height": "50px",
                                                "width": "50px",
                                              }
                }
/>,
              <p />,
            ],
            "id": undefined,
            "style": Object {
              "height": "100px",
              "margin": "20px",
              "width": "100px",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "",
                "src": "undefined",
                "style": Object {
                  "borderRadius": "50px",
                  "height": "50px",
                  "width": "50px",
                },
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "p",
            },
          ],
          "type": "li",
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
